OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,401 @@
+                    $"ODP event send failed (queueSize = {_queue.Count}).");
+                return;
+            }
+
+            _queue.Enqueue(odpEvent);
+
+            ProcessQueue();
+        }
+
+        /// <summary>
+        /// Process the queue
+        /// </summary>
+        /// <param name="shouldFlush">True if complete flush of queue is needed</param>
+        private void ProcessQueue(bool shouldFlush = false)
+        {
+            if (State != ExecutionState.Running)
+            {
+                return;
+            }
+
+            if (!IsOdpConfigurationReady())
+            {
+                return;
+            }
+
+            _logger.Log(LogLevel.DEBUG,
+                $"Processing Queue {(shouldFlush ? "(flush)" : string.Empty)}");
+
+            if (shouldFlush)
+            {
+                ClearCurrentTimeout();
+
+                State = ExecutionState.Processing;
+
+                while (QueueContainsItems())
+                {
+                    MakeAndSend1Batch();
+                }
+            }
+            else if (QueueHasBatches())
+            {
+                ClearCurrentTimeout();
+
+                State = ExecutionState.Processing;
+
+                while (QueueHasBatches())
+                {
+                    MakeAndSend1Batch();
+                }
+            }
+
+            State = ExecutionState.Running;
+            SetNewTimeout();
+        }
+
+        /// <summary>
+        /// Make a single batch and send it
+        /// </summary>
+        private void MakeAndSend1Batch()
+        {
+            var batch = new List<OdpEvent>(_batchSize);
+
+            // remove a batch from the queue
+            for (int i = 0; i < _batchSize && _queue.Count > 0; i += 1)
+            {
+                if (_queue.TryDequeue(out OdpEvent dequeuedOdpEvent))
+                {
+                    batch.Add(dequeuedOdpEvent);
+                }
+            }
+
+            if (batch.Count > 0)
+            {
+                Task.Run(() =>
+                {
+                    var shouldRetry = false;
+                    var attemptNumber = 0;
+                    do
+                    {
+                        shouldRetry = _odpEventApiManager.SendEvents(_odpConfig.ApiKey,
+                            _odpConfig.ApiHost, batch);
+                        attemptNumber += 1;
+                    } while (shouldRetry && attemptNumber < MAX_RETRIES);
+                });
+            }
+        }
+
+        /// <summary>
+        /// Start a new timer to begin the next flush interval
+        /// </summary>
+        private void SetNewTimeout()
+        {
+            _timeoutToken = new CancellationTokenSource();
+            var ct = _timeoutToken.Token;
+            Task.Run(() =>
+            {
+                Thread.Sleep(_flushInterval);
+                if (!ct.IsCancellationRequested)
+                    ProcessQueue(true);
+            }, ct);
```

Comment:
@msohailhussain I think some of the inconsistencies during unit test execution stem from this `Task.Run`, which as I understand it, are executed on from the same thread pool and there are no guarantees of separation.